### PR TITLE
:book: Add kubeadmin igintion generator to list

### PR DESCRIPTION
**Pull request by MartinForReal**
Signed-off-by: Fan Shang Xiang <fanshangxiang@gmail.com>
**What this PR does / why we need it**:

Add kubeadmin bootstrap generator for ignition to providers list.

### CODE REVIEW

**Comment by fabriziopandini** [comment]
it is awesome to see the community extending the Cluster API ecosystem and being able to use the existing extensibility points! Great!

File: docs/book/src/reference/providers.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ support v1alpha3 API types.
 
 ## Bootstrap
 - [**Kubeadm**](https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm)
+- [**Kubeadm with ignition on AWS**](https://github/minsheng-fintech-corp-ltd/cluster-api-bootstrap-provider-kubeadm-ignition)
```
Comment (ncdc) [inline docs/book/src/reference/providers.md:10]:
```suggestion
- [**Kubeadm with ignition on AWS**](https://github.com/minsheng-fintech-corp-ltd/cluster-api-bootstrap-provider-kubeadm-ignition)
```

**Comment by detiber** [comment]
@MartinForReal From what I can tell by the provided link is that this controller is meant to replace the existing bootstrap controller but re-uses the same types. This is not necessarily a bad thing, but it does limit the potential use cases. 

We do have an outstanding issue for the in-tree kubeadm bootstrap provider to support multiple "formats" (https://github.com/kubernetes-sigs/cluster-api/issues/1582) which we might be able to leverage to fold this functionality into the existing in-tree controller.

Alternatively, having separate types for cluster-api-bootstrap-provider-kubeadm-ignition would allow for co-existence with the in-tree bootstrap provider.

**Comment by neolit123** [comment]
the https://github.com/minsheng-fintech-corp-ltd/cluster-api-bootstrap-provider-kubeadm-ignition readme needs an update first to make it unique to the new BS provider (currently == CAPBK).

possibly add this new provider here after that?

**Comment by ncdc** [comment]
Continuing on Jason's comments, if you're willing, it would be great to explore:

1. Does it make sense to pull your ignition code into CABPK (separate `format`)?
1. If not, is there anything we could do to minimize code duplication in your repo by making CABPK better structured as a library (bootstrap token generation, etc.)?

File: docs/book/src/reference/providers.md (hunk 1 above)
Comment (MartinForReal) [reply docs/book/src/reference/providers.md]: I've updated the link.Thanks!

**Comment by MartinForReal** [comment]
> the https://github.com/minsheng-fintech-corp-ltd/cluster-api-bootstrap-provider-kubeadm-ignition readme needs an update first to make it unique to the new BS provider (currently == CAPBK).
> 
> possibly add this new provider here after that?

I will update the project's readme file, reference, and other documents and leave comments when ready.

**Comment by MartinForReal** [comment]
> Continuing on Jason's comments, if you're willing, it would be great to explore:
> 
> 1. Does it make sense to pull your ignition code into CABPK (separate `format`)?
> 2. If not, is there anything we could do to minimize code duplication in your repo by making CABPK better structured as a library (bootstrap token generation, etc.)?

I don't think putting this provider into CABPK is a good idea because it heavily depends on the template [here](https://github.com/minsheng-fintech-corp-ltd/k8s-artifact-auto-sync) and artifacts in s3 in order to address 64KB limit of user data on AWS. Ignition doesn't have support for AWS secret manager as cloud-init does. The implementation is a bit tricky but it works. 

In the short run, I think the bootstrap user data generator needs Kubeadm config file, certificates, and other files(audit policy, postscripts and etc) and be invoked when the CABPK is generating user data for the first node or newly-joined node. If we can provide some utility functions to generate a script file that executes kubeadm command, the generator will only need to care if the files exist before the scripts are executed.  So I think a GRPC protocol and some utility functions would be enough. 

In the long run, I think this problem is similar to the one in volume binding scenario. Volumes should be provisioned based on the scheduler's decision. In our scenario, if we add an extra step called "Scheduled" in the machine's lifecycle, then we can implement bootstrap provider more easily because we can get more info about infrastructure(such as subnet, available zone) and choose the image as needed.  So if we will implement a generator in this way, we need to add a step in the machine's lifecycle and let the infrastructure provider decide how to generate user data.

**Comment by ncdc** [comment]
>  If we can provide some utility functions to generate a script file that executes kubeadm command, the generator will only need to care if the files exist before the scripts are executed. So I think a GRPC protocol and some utility functions would be enough.

I'm not sure how GRPC fits into this?

**Comment by MartinForReal** [comment]
> I'm not sure how GRPC fits into this?

AFAIK, It's a possible way to implement RPC. If we can create more CR resources. That would be fine too. The userdata is only used once and should be destroyed after being used. So I think the provider implements grpc service and then CAPBK can add these service instances into memory through load config from configmap and consume these services using grpc client. But I can hardly agree with the idea that multiple controllers watch the same type of object because it is very hard to keep the object's state up-to-date.

**Comment by ncdc** [comment]
Are you saying you'd have CABPK ask some other "generator service" to generate the bootstrap data, and you'd do this via either gRPC or CRDs?

**Comment by MartinForReal** [comment]
> Are you saying you'd have CABPK ask some other "generator service" to generate the bootstrap data, and you'd do this via either gRPC or CRDs?

Yes. Exactly!  Sorry for my poor written English.

**Comment by vincepri** [comment]
@MartinForReal Are you still interested in getting this PR in?

**Comment by vincepri** [comment]
Closing this for now given the lack of updates, @MartinForReal feel free to reopen when you're ready!

/close

**Pull request closed without merging**